### ISSUE TITLE
Using Embedded Jetty with JSPs and Spring Boot

### ISSUE THREAD

**Issue by ghillert** [eclipse/jetty.project#420]
Currently, as of Spring Boot version `1.3.3` it is not possible to use Jetty with JSPs in embedded mode with Spring Boot. 

Relevant Issues: https://github.com/spring-projects/spring-boot/issues/367

Some of the issues, I have seen, can be worked around but a major issue seems the loading of the JSP Tag libraries. I got it to work inside the IDE (STS) but once you run the application using just the über-jar, things fail:

```
org.apache.jasper.JasperException: /welcome.jsp(3,68) PWC6132: Unable to read TLD "META-INF/spring.tld" from JAR file "jar:file:/Users/hillert/dev/git/spring-boot-jsp-demo/target/jsp-demo-0.0.1-SNAPSHOT.jar!/BOOT-INF/lib/spring-webmvc-4.3.0.BUILD-SNAPSHOT.jar": java.io.IOException: Unable to open root Jar file 'jar:file:/Users/hillert/dev/git/spring-boot-jsp-demo/target/jsp-demo-0.0.1-SNAPSHOT.jar'
``` 

Just for reference, there is also a PR for executable wars - https://github.com/spring-projects/spring-boot/pull/5290

Just FYI, there might be also other improvements that could be made - e.g. it looks like Jetty by default does not handle Servlet 3.0 external resources. Customization as described here was necessary:

https://www.totaalsoftware.com/blog/104-jetty-with-servlet-3-0-external-resources

Lastly, Tomcat is not perfect by any means in regards to serving JSP via über-Jars, but it works by either creating an executable war-file or by adding your JSPs to `META-INF/resource` when using executable uber-jars:

https://github.com/ghillert/spring-boot-jsp-demo

**Comment by joakime** [comment]
I can't find a use of [`JettyJspServlet`](http://download.eclipse.org/jetty/9.2.15.v20160210/apidocs/org/eclipse/jetty/jsp/JettyJspServlet.html) in the [spring-project/spring-boot](https://github.com/spring-projects/spring-boot/search?utf8=%E2%9C%93&q=JettyJspServlet) project.

See: [jetty-project/embedded-jetty-jsp](https://github.com/jetty-project/embedded-jetty-jsp) for [example](https://github.com/jetty-project/embedded-jetty-jsp/blob/master/src/main/java/org/eclipse/jetty/demo/Main.java#L203)

Nor, do I see the [classloader requirements for JSTL](https://github.com/jetty-project/embedded-jetty-jsp/blob/master/src/main/java/org/eclipse/jetty/demo/Main.java#L186-L196) being used in the spring-boot project.

**Comment by tsachev** [comment]
currently spring boot will try to register the jasper's `org.apache.jasper.servlet.JspServlet`, but this can be overridden with a property `server.jsp-servlet.class-name`. In my PR (spring-projects/spring-boot#5290) I also create poor man's url stream handler for "war" protocol. Latter I realized that this is also supported by jetty's module `jetty-osgi-boot-warurl` but I really do not want to involve osgi here. Probably the `WarUrlStreamHandler` can be moved in a more common place.

Spring boot has its own class loader as part of the `spring-boot-loader` module which I think works ok.
There is a problem with spring-boot-loader and jetty 9.3 (caused be using path instead of file internally) which I already reported in spring-boot spring-projects/spring-boot#5287 with PR for it spring-projects/spring-boot##5289.

I wonder what @wilkinsona and/or @philwebb think?

**Comment by janbartel** [comment]
@ghillert do you have an example project for this? Jetty by default certainly does support META-INF/resources in jars that are considered to be on the webapp's classpath. 

Jan

**Comment by ghillert** [comment]
@janbartel My Jetty sample can be found here: https://github.com/ghillert/spring-boot-jsp-demo/tree/jetty

@joakime I am setting the `JettyJspServlet` at https://github.com/ghillert/spring-boot-jsp-demo/blob/jetty/src/main/resources/application.properties#L3 and I also set the ClassLoader here:

https://github.com/ghillert/spring-boot-jsp-demo/blob/jetty/src/main/java/com/hillert/JspDemoApplication.java#L80

The application works inside my IDE (Eclipse). However, when executing the Jar, I receive the following error:

```
2016-03-11 15:21:21.598  INFO 64007 --- [           main] com.hillert.JspDemoApplication           : Started JspDemoApplication in 8.389 seconds (JVM running for 9.436)
2016-03-11 15:21:31.743  WARN 64007 --- [tp1959698446-17] o.eclipse.jetty.servlet.ServletHandler   :

org.apache.jasper.JasperException: /WEB-INF/jsp/hello-world.jsp(6,64) PWC6132: Unable to read TLD "META-INF/c.tld" from JAR file "jar:file:/Users/hillert/dev/git/spring-boot-jsp-demo/target/jsp-demo-0.0.1-SNAPSHOT.jar!/lib/javax.servlet.jsp.jstl-1.2.2.jar": java.io.IOException: Unable to open root Jar file 'jar:file:/Users/hillert/dev/git/spring-boot-jsp-demo/target/jsp-demo-0.0.1-SNAPSHOT.jar'
	at org.apache.jasper.compiler.DefaultErrorHandler.jspError(DefaultErrorHandler.java:92) ~[javax.servlet.jsp-2.3.2.jar!/:2.3.2]
	at org.apache.jasper.compiler.ErrorDispatcher.dispatch(ErrorDispatcher.java:378) ~[javax.servlet.jsp-2.3.2.jar!/:2.3.2]
	at org.apache.jasper.compiler.ErrorDispatcher.jspError(ErrorDispatcher.java:172) ~[javax.servlet.jsp-2.3.2.jar!/:2.3.2]
	at org.apache.jasper.compiler.TagLibraryInfoImpl.<init>(TagLibraryInfoImpl.java:273) ~[javax.servlet.jsp-2.3.2.jar!/:2.3.2]
	at org.apache.jasper.compiler.Parser.parseTaglibDirective(Parser.java:502) ~[javax.servlet.jsp-2.3.2.jar!/:2.3.2]
	at org.apache.jasper.compiler.Parser.parseDirective(Parser.java:582) ~[javax.servlet.jsp-2.3.2.jar!/:2.3.2]
	at org.apache.jasper.compiler.Parser.parseElements(Parser.java:1657) ~[javax.servlet.jsp-2.3.2.jar!/:2.3.2]
	at org.apache.jasper.compiler.Parser.parse(Parser.java:185) ~[javax.servlet.jsp-2.3.2.jar!/:2.3.2]
	at org.apache.jasper.compiler.ParserController.doParse(ParserController.java:244) ~[javax.servlet.jsp-2.3.2.jar!/:2.3.2]
	at org.apache.jasper.compiler.ParserController.parse(ParserController.java:145) ~[javax.servlet.jsp-2.3.2.jar!/:2.3.2]
	at org.apache.jasper.compiler.Compiler.generateJava(Compiler.java:212) ~[javax.servlet.jsp-2.3.2.jar!/:2.3.2]
	at org.apache.jasper.compiler.Compiler.compile(Compiler.java:451) ~[javax.servlet.jsp-2.3.2.jar!/:2.3.2]
	at org.apache.jasper.JspCompilationContext.compile(JspCompilationContext.java:625) ~[javax.servlet.jsp-2.3.2.jar!/:2.3.2]
	at org.apache.jasper.servlet.JspServletWrapper.service(JspServletWrapper.java:375) ~[javax.servlet.jsp-2.3.2.jar!/:2.3.2]
	at org.apache.jasper.servlet.JspServlet.serviceJspFile(JspServlet.java:473) ~[javax.servlet.jsp-2.3.2.jar!/:2.3.2]
	at org.apache.jasper.servlet.JspServlet.service(JspServlet.java:377) ~[javax.servlet.jsp-2.3.2.jar!/:2.3.2]
	at org.eclipse.jetty.jsp.JettyJspServlet.service(JettyJspServlet.java:103) ~[jetty-jsp-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790) ~[javax.servlet-api-3.1.0.jar!/:3.1.0]
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:812) ~[jetty-servlet-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1669) ~[jetty-servlet-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:101) ~[spring-web-4.2.5.RELEASE.jar!/:4.2.5.RELEASE]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652) ~[jetty-servlet-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:101) ~[spring-web-4.2.5.RELEASE.jar!/:4.2.5.RELEASE]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652) ~[jetty-servlet-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:101) ~[spring-web-4.2.5.RELEASE.jar!/:4.2.5.RELEASE]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652) ~[jetty-servlet-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:101) ~[spring-web-4.2.5.RELEASE.jar!/:4.2.5.RELEASE]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652) ~[jetty-servlet-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:585) [jetty-servlet-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143) [jetty-server-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:595) [jetty-security-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:223) [jetty-server-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127) [jetty-server-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515) [jetty-servlet-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185) [jetty-server-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061) [jetty-server-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141) [jetty-server-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.eclipse.jetty.server.Dispatcher.forward(Dispatcher.java:191) ~[jetty-server-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.eclipse.jetty.server.Dispatcher.forward(Dispatcher.java:72) ~[jetty-server-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.springframework.web.servlet.view.InternalResourceView.renderMergedOutputModel(InternalResourceView.java:168) ~[spring-webmvc-4.2.5.RELEASE.jar!/:4.2.5.RELEASE]
	at org.springframework.web.servlet.view.AbstractView.render(AbstractView.java:303) ~[spring-webmvc-4.2.5.RELEASE.jar!/:4.2.5.RELEASE]
	at org.springframework.web.servlet.DispatcherServlet.render(DispatcherServlet.java:1243) ~[spring-webmvc-4.2.5.RELEASE.jar!/:4.2.5.RELEASE]
	at org.springframework.web.servlet.DispatcherServlet.processDispatchResult(DispatcherServlet.java:1027) ~[spring-webmvc-4.2.5.RELEASE.jar!/:4.2.5.RELEASE]
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:971) ~[spring-webmvc-4.2.5.RELEASE.jar!/:4.2.5.RELEASE]
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:893) ~[spring-webmvc-4.2.5.RELEASE.jar!/:4.2.5.RELEASE]
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:968) ~[spring-webmvc-4.2.5.RELEASE.jar!/:4.2.5.RELEASE]
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:859) ~[spring-webmvc-4.2.5.RELEASE.jar!/:4.2.5.RELEASE]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:687) ~[javax.servlet-api-3.1.0.jar!/:3.1.0]
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:844) ~[spring-webmvc-4.2.5.RELEASE.jar!/:4.2.5.RELEASE]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790) ~[javax.servlet-api-3.1.0.jar!/:3.1.0]
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:812) ~[jetty-servlet-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1669) ~[jetty-servlet-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.eclipse.jetty.websocket.server.WebSocketUpgradeFilter.doFilter(WebSocketUpgradeFilter.java:224) ~[websocket-server-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652) ~[jetty-servlet-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:99) ~[spring-web-4.2.5.RELEASE.jar!/:4.2.5.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) ~[spring-web-4.2.5.RELEASE.jar!/:4.2.5.RELEASE]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652) ~[jetty-servlet-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.springframework.web.filter.HttpPutFormContentFilter.doFilterInternal(HttpPutFormContentFilter.java:87) ~[spring-web-4.2.5.RELEASE.jar!/:4.2.5.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) ~[spring-web-4.2.5.RELEASE.jar!/:4.2.5.RELEASE]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652) ~[jetty-servlet-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.springframework.web.filter.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:77) ~[spring-web-4.2.5.RELEASE.jar!/:4.2.5.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) ~[spring-web-4.2.5.RELEASE.jar!/:4.2.5.RELEASE]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652) ~[jetty-servlet-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:121) ~[spring-web-4.2.5.RELEASE.jar!/:4.2.5.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) ~[spring-web-4.2.5.RELEASE.jar!/:4.2.5.RELEASE]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652) ~[jetty-servlet-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:585) [jetty-servlet-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143) [jetty-server-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:577) [jetty-security-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:223) [jetty-server-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127) [jetty-server-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515) [jetty-servlet-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185) [jetty-server-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061) [jetty-server-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141) [jetty-server-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97) [jetty-server-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.eclipse.jetty.server.Server.handle(Server.java:499) [jetty-server-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:311) [jetty-server-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257) [jetty-server-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544) [jetty-io-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635) [jetty-util-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555) [jetty-util-9.2.15.v20160210.jar!/:9.2.15.v20160210]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_60]
```

**Comment by ghillert** [comment]
@janbartel Regarding `META-INF/resources` - I had to do this:

```
ClassPathResource classPathResource = new ClassPathResource("META-INF/resources");
					String externalResource = classPathResource.getURI().toString();
					String[] resources = new String[] { externalResource };
					webAppContext.setBaseResource(new ResourceCollection(resources));
```

See: https://github.com/ghillert/spring-boot-jsp-demo/blob/jetty/src/main/java/com/hillert/JspDemoApplication.java#L78

Adapted from: https://www.totaalsoftware.com/blog/104-jetty-with-servlet-3-0-external-resources

If I leave it out, I get:

```
2016-03-11 15:46:40.610 ERROR 66840 --- [qtp194408994-23] org.apache.jasper.servlet.JspServlet     : PWC6117: File "/private/var/folders/hs/h87zy7z17qs6mcnl4hj8_dp00000gp/T/jetty-docbase.3309716627247661555.8080/WEB-INF/jsp/hello-world.jsp" not found
```

So it is interesting that `META-INF/resources` is not served by default.

**Comment by ghillert** [comment]
An interesting observation regarding the TLD parsing issue as mentioned earlier:

```
 PWC6132: Unable to read TLD "META-INF/c.tld" from JAR file
```

I wonder if this is somewhat similar to what I experience with Undertow - the need to retrieve the TLDs manually. As executable Jars in Spring Boot contain nested jar, this may be the issue. Is there a way to pass in the TLDs information `manually` with Jetty?

In Spring we have the `PathMatchingResourcePatternResolver` class that is able to retrieve classpath resources from nested Jars:

```
		final ResourcePatternResolver resolver = new PathMatchingResourcePatternResolver(loader);
		final Resource[] resources;
		final String locationPattern = "classpath*:**/*.tld";
		try {
			resources = resolver.getResources(locationPattern);
		} catch (IOException e) {
			throw new IllegalStateException(String.format("Error while retrieving resources"
					+ "for location pattern '%s'.", locationPattern, e));
		}
```

From: 

https://github.com/ghillert/spring-boot-jsp-demo/blob/undertow/src/main/java/com/hillert/TldLocator.java

**Comment by joakime** [comment]
Just tossing in some information to this thread ...

Jetty scans for and adds references to the `META-INF/resources` directories during the execution of [`MetaInfConfiguration`](https://github.com/eclipse/jetty.project/blob/jetty-9.3.8.RC0/jetty-webapp/src/main/java/org/eclipse/jetty/webapp/MetaInfConfiguration.java#L156-L218).

The `MetaInfConfiguration` will execute as part of a configured `WebAppContext`

> Note: These `Configurations` are not available if you are using `ServletContextHandler`

Also note, the unpacking of a `WebAppContext` can result in each of those `META-INF/reosurces` directories existing in the context's temp directory (and being served from that unpacked temp directory location)

**Comment by janbartel** [comment]
@joakime I think the problem is that in spring-boot, they do not use **any** of our normal Configuration classes associated with a WebAppContext. Thus, as MetaInfConfiguration is not executed, we do not look for META-INF/resources, nor META-INF/*.tlds nor META-INF/web-fragment.xml (although I guess the latter is not really an issue for spring-boot as I gather the purpose is to be web.xml-less).

**Comment by janbartel** [comment]
I've had a better look at pull request https://github.com/spring-projects/spring-boot/pull/5290, which is a nifty piece of work from @tsachev. Jsps and jstl work correctly with the patch, and I've also tested using dependency jars containing META-INF/resources jsps and they are all also correctly served. I've also tested using jsf with this patch in place, and it is also working correctly.

**Comment by wilkinsona** [comment]
@janbartel Thank you

**Comment by janbartel** [comment]
Note that we've released jetty-9.2.16, which updates the version of jasper, and helps with @tsachev's pull request which makes jsp/jstl work correctly. I'll close this issue now.
